### PR TITLE
Clamp NaN to white when converting to integer

### PR DESCRIPTION
**Pull request by FlareFlo**
Linked issue: https://github.com/image-rs/image/issues/2275

### CODE REVIEW

**Comment by FlareFlo** [comment]
No clue why clippy throws a fit, this PR does not touch those failing files at all.

**Comment by FlareFlo** [comment]
> There's new clippy complaints every time a new clippy version comes out. Don't worry about them.
> 
> Do you know the performance implications of this change? I'm not sure performance was great to start with, but I don't want to unintentionally add guarantees that will make it unnecessary slow forever.

https://godbolt.org/z/7318KPd4f Not great, but im not sure what else to do

File: src/color.rs [hunk 1]
```diff
@@ -385,17 +385,29 @@ impl<T: Primitive> FromPrimitive<T> for T {
...
+    } else {
+        i
+    }
+}
+
 impl FromPrimitive<f32> for u8 {
     fn from_primitive(float: f32) -> Self {
         let inner = (float.clamp(0.0, 1.0) * u8::MAX as f32).round();
-        NumCast::from(inner).unwrap()
+        NumCast::from(sanitize_nan(inner)).unwrap()
```
Comment (fintelia) [inline src/color.rs:403]: Wait, shouldn't the `sanitize_nan` call happen before scaling by `u8::MAX`?

Comment (FlareFlo) [reply src/color.rs:403]: Good point, ill fix that in a sec

**Review by fintelia — APPROVED** [review]

**Comment by kornelski** [comment]
It should be possible to replace `clamp` with custom comparison that handles `NaN` at zero cost:

`if !(x < 1.0) { 1.0 }` handles both `NaN` and `>1.0`.

**Comment by FlareFlo** [comment]
> It should be possible to replace `clamp` with custom comparison that handles `NaN` at zero cost:
> 
> `if !(x < 1.0) { 1.0 }` handles both `NaN` and `>1.0`.

Something like this appears to produce better codegen indeed https://godbolt.org/z/qnbjWhr3c

**Comment by FlareFlo** [comment]
> There's new clippy complaints every time a new clippy version comes out. Don't worry about them.
> 
> Do you know the performance implications of this change? I'm not sure performance was great to start with, but I don't want to unintentionally add guarantees that will make it unnecessary slow forever.

I have a few ideas for speeding things up, too much for this PR though. Ill experiment and open a PR when i have something that might make sense.

**Review by fintelia — APPROVED** [review]

**Pull request merged**